usr/man/man/usr/share/man/man1/ckitem.1
'\" te
.\"  Copyright 1989 AT&T  Copyright (c) 1992, Sun Microsystems, Inc.  All Rights Reserved
.TH ckitem 1 "14 Sep 1992" "SunOS 5.11" "User Commands"
.SH NAME
ckitem, erritem, helpitem \- build a menu; prompt for and return a menu item
.SH SYNOPSIS
.LP
.nf
\fBckitem\fR [\fB-Q\fR] [\fB-W\fR \fIwidth\fR] [\fB-uno\fR] [\fB-f\fR \fIfilename\fR] [\fB-l\fR \fIlabel\fR] 
     [ [\fB-i\fR \fIinvis\fR] [,]...] [\fB-m\fR \fImax\fR] [\fB-d\fR \fIdefault\fR] [\fB-h\fR \fIhelp\fR] 
     [\fB-e\fR \fI error\fR] [\fB-p\fR \fIprompt\fR] [\fB-k\fR \fIpid\fR [\fB-s\fR \fIsignal\fR]] 
     [\fIchoice\fR [...]]
.fi

.LP
.nf
\fB/usr/sadm/bin/erritem\fR [\fB-W\fR \fIwidth\fR] [\fB-e\fR \fIerror\fR] [\fIchoice\fR [..]]
.fi

.LP
.nf
\fB/usr/sadm/bin/helpitem\fR [\fB-W\fR \fIwidth\fR] [\fB-h\fR \fIhelp\fR] [\fIchoice\fR [..]]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBckitem\fR utility builds a menu and prompts the user to choose one item from a menu of items. It then  verifies the response. Options for this command define, among other things, a prompt message whose response will be a menu item, text for help and error messages, and a default value (which will be returned if the user responds with a carriage return).
.sp
.LP
By default, the menu is formatted so that each item is prepended by a number and is printed in columns across the terminal. Column length is determined by the longest choice. Items are alphabetized.
.sp
.LP
All messages are limited in length to 70 characters and are formatted automatically. Any white space used in the definition (including newline) is stripped. The \fB-W\fR option cancels the automatic formatting. When a tilde is placed at the beginning or end of a message definition, the default text will be inserted at that point, allowing both custom text and the default text to be displayed.
.sp
.LP
If the prompt, help or error message is not defined, the default message (as defined under \fBNOTES\fR) will be displayed.
.sp
.LP
Two visual tool modules are linked to the \fBckitem\fR command. They are \fBerritem\fR (which formats and displays an error message) and \fBhelpitem\fR (which formats and displays a help message). These modules should be used in conjunction with \fBFML\fR objects. In this instance, the \fBFML\fR object defines the prompt. When \fIchoice\fR is defined in these modules, the messages will describe the available menu choice (or choices).
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB-d\fR\fI default\fR\fR
.ad
.RS 15n
.rt  
Define the default value as \fIdefault\fR. The default is not validated and so does not have to meet any criteria.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-e\fR\fI error\fR\fR
.ad
.RS 15n
.rt  
Define the error message as \fI error\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-f\fR\fI filename\fR\fR
.ad
.RS 15n
.rt  
Define a file, \fIfilename\fR, which contains a list of menu items to be displayed. (The format of this file is: \fBtoken<tab>description\fR. Lines beginning with a pound sign (#) are designated as comments and ignored.)
.RE

.sp
.ne 2
.mk
.na
\fB\fB-h\fR\fI help\fR\fR
.ad
.RS 15n
.rt  
Define the help messages as \fI help\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-i\fR\fI invis\fR\fR
.ad
.RS 15n
.rt  
Define invisible menu choices (those which will not be printed in the menu). (For example, ``all'' used as an invisible choice would mean it is a legal option but does not appear in the menu. Any number of invisible choices may be defined.) Invisible choices should be made known to a user either in the prompt or in a help message.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-k\fR\fI pid\fR\fR
.ad
.RS 15n
.rt  
Specify that the process \fBID\fR \fIpid\fR is to be sent a signal if the user chooses to abort.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-l\fR\fI label\fR\fR
.ad
.RS 15n
.rt  
Define a label, \fIlabel\fR, to print above the menu.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-m\fR\fI max\fR\fR
.ad
.RS 15n
.rt  
Define the maximum number of menu choices that the user can choose. The default is 1.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-n\fR\fR
.ad
.RS 15n
.rt  
Specify that menu items should not be displayed in alphabetical order.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-o\fR\fR
.ad
.RS 15n
.rt  
Specify that only one menu token will be returned.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-p\fR\fI prompt\fR\fR
.ad
.RS 15n
.rt  
Define the prompt message as \fIprompt\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-Q\fR\fR
.ad
.RS 15n
.rt  
Specify that quit will not be allowed as a valid response.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-s\fR\fI signal\fR\fR
.ad
.RS 15n
.rt  
Specify that process \fBID\fR \fIpid\fR defined with the \fB-k\fR option is to be sent signal \fI signal\fR when quit is chosen. If no signal is specified, \fBSIGTERM\fR is used.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-u\fR\fR
.ad
.RS 15n
.rt  
Specify that menu items should be displayed as an unnumbered list.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-W\fR\fI width\fR\fR
.ad
.RS 15n
.rt  
Specify that prompt, help and error messages will be formatted to a line length of \fIwidth\fR.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.mk
.na
\fB\fIchoice\fR\fR
.ad
.RS 10n
.rt  
Define menu items. Items should be separated by white space or newline.
.RE

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.mk
.na
\fB\fB0\fR\fR
.ad
.RS 5n
.rt  
Successful execution.
.RE

.sp
.ne 2
.mk
.na
\fB\fB1\fR\fR
.ad
.RS 5n
.rt  
\fBEOF\fR on input, or negative width on \fB-W\fR option, or inability to open file on \fB-f\fR option, or usage error.
.RE

.sp
.ne 2
.mk
.na
\fB\fB3\fR\fR
.ad
.RS 5n
.rt  
User termination (quit).
.RE

.sp
.ne 2
.mk
.na
\fB\fB4\fR\fR
.ad
.RS 5n
.rt  
No choices from which to choose.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
.TE

.SH SEE ALSO
.sp
.LP
\fBattributes\fR(5)
.SH NOTES
.sp
.LP
The user may input the number of the menu item if choices are numbered or as much of the string required for a unique identification of the item. Long menus are paged with 10 items per page.
.sp
.LP
When menu entries are defined both in a file (by using the \fB-f\fR option) and also on the command line, they are usually combined alphabetically. However, if the \fB-n\fR option is used to suppress  alphabetical ordering, then the entries defined in the file are shown first, followed by the options defined on the command line.
.sp
.LP
The default prompt for \fBckitem\fR is:
.sp
.in +2
.nf
Enter selection [?,??,q]:
.fi
.in -2
.sp

.sp
.LP
One question mark will give a help message and then redisplay the prompt. Two question marks will give a help message and then redisplay the menu label, the menu and the prompt.
.sp
.LP
The default error message if you typed a number is:
.sp
.in +2
.nf
ERROR: Bad numeric choice specification
.fi
.in -2
.sp

.sp
.LP
The default error message if you typed a string is:
.sp
.in +2
.nf
ERROR: Entry does not match available menu selection. Enter the number
of the menu item you wish to select, the token which is associated
with the menu item, or a partial string which uniquely identifies the
token for the menu item. Enter ?? to reprint the menu.
.fi
.in -2
.sp

.sp
.LP
The default help message is:
.sp
.in +2
.nf
Enter the number of the menu item you wish to select, the token
which is associated with the menu item, or a partial string which
uniquely identifies the token for the menu item. Enter ? to
reprint the menu.
.fi
.in -2
.sp

.sp
.LP
When the quit option is chosen (and allowed), \fBq\fR is returned along with the return code \fB3\fR.


usr/src/lib/libadm/common/ckitem.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright (c) 1989, 2011, Oracle and/or its affiliates. All rights reserved.
 */

#include <stdio.h>
#include <ctype.h>
#include <limits.h>
#include "valtools.h"
#include <sys/types.h>
#include <stdlib.h>
#include <strings.h>
#include "libadm.h"

static int	insert(struct _choice_ *, CKMENU *);
static char	*strtoki(char *, char *);
static char	**match(CKMENU *, char *, int);
static int	getstr(char *, char *, char *, char *, char *);
static int	getnum(char *, int, int *, int *);
static struct _choice_ *next(struct _choice_ *);

static char	*deferr;
static char	*errmsg;
static char	*defhlp;

#define	PROMPT	"Enter selection"
#define	MESG0	"Entry does not match available menu selection. "
#define	MESG1	"the number of the menu item you wish to select, or "
#define	MESG2	"the token which is associated with the menu item,\
		or a partial string which uniquely identifies the \
		token for the menu item. Enter ?? to reprint the menu."

#define	TOOMANY	"Too many items selected from menu"
#define	NOTUNIQ	"The entered text does not uniquely identify a menu choice."
#define	BADNUM	"Bad numeric choice specification"

static char *
setmsg(CKMENU *menup, short flag)
{
	int	n;
	char	*msg;

	n = (int)(6 + sizeof (MESG2));
	if (flag)
		n += (int)(sizeof (MESG0));

	if (menup->attr & CKUNNUM) {
		msg = calloc((size_t)n, sizeof (char));
		if (flag)
			(void) strcpy(msg, MESG0);
		else
			msg[0] = '\0';
		(void) strcat(msg, "Enter ");
		(void) strcat(msg, MESG2);
	} else {
		msg = calloc(n+sizeof (MESG1), sizeof (char));
		if (flag)
			(void) strcpy(msg, MESG0);
		else
			msg[0] = '\0';
		(void) strcat(msg, "Enter ");
		(void) strcat(msg, MESG1);
		(void) strcat(msg, MESG2);
	}
	return (msg);
}

CKMENU *
allocmenu(char *label, int attr)
{
	CKMENU *pt;

	if (pt = calloc(1, sizeof (CKMENU))) {
		pt->attr = attr;
		pt->label = label;
	}
	return (pt);
}

void
ckitem_err(CKMENU *menup, char *error)
{
	deferr = setmsg(menup, 1);
	puterror(stdout, deferr, error);
	free(deferr);
}

void
ckitem_hlp(CKMENU *menup, char *help)
{
	defhlp = setmsg(menup, 0);
	puthelp(stdout, defhlp, help);
	free(defhlp);
}

int
ckitem(CKMENU *menup, char *item[], short max, char *defstr, char *error,
	char *help, char *prompt)
{
	int	n, i;
	char	strval[MAX_INPUT];
	char	**list;

	if ((menup->nchoices <= 0) && !menup->invis)
		return (4); /* nothing to choose from */

	if (menup->attr & CKONEFLAG) {
		if (((n = menup->nchoices) <= 1) && menup->invis) {
			for (i = 0; menup->invis[i]; ++i)
				n++;
		}
		if (n <= 1) {
			if (menup->choice)
				item[0] = menup->choice->token;
			else if (menup->invis)
				item[0] = menup->invis[0];
			item[1] = NULL;
			return (0);
		}
	}

	if (max < 1)
		max = menup->nchoices;

	if (!prompt)
		prompt = PROMPT;
	defhlp = setmsg(menup, 0);
	deferr = setmsg(menup, 1);

reprint:
	printmenu(menup);

start:
	if (n = getstr(strval, defstr, error, help, prompt)) {
		free(defhlp);
		free(deferr);
		return (n);
	}
	if (strcmp(strval, "??") == 0) {
		goto reprint;
	}
	if ((defstr) && (strcmp(strval, defstr) == 0)) {
		item[0] = defstr;
		item[1] = NULL;
	} else {
		list = match(menup, strval, (int)max);
		if (!list) {
			puterror(stderr, deferr, (errmsg ? errmsg : error));
			goto start;
		}
		for (i = 0; (i < max); i++)
			item[i] = list[i];
		free(list);
	}
	free(defhlp);
	free(deferr);
	return (0);
}

static int
getnum(char *strval, int max, int *begin, int *end)
{
	int n;
	char *pt;

	*begin = *end = 0;
	pt = strval;
	for (;;) {
		if (*pt == '$') {
			n = max;
			pt++;
		} else {
			n = (int)strtol(pt, &pt, 10);
			if ((n <= 0) || (n > max))
				return (1);
		}
		while (isspace((unsigned char)*pt))
			pt++;

		if (!*begin && (*pt == '-')) {
			*begin = n;
			pt++;
			while (isspace((unsigned char)*pt))
				pt++;
			continue;
		} else if (*pt) {
			return (1); /* wasn't a number, or an invalid one */
		} else if (*begin) {
			*end = n;
			break;
		} else {
			*begin = n;
			break;
		}
	}
	if (!*end)
		*end = *begin;
	return ((*begin <= *end) ? 0 : 1);
}

static char **
match(CKMENU *menup, char *strval, int max)
{
	struct _choice_ *chp;
	char **choice;
	int begin, end;
	char *pt, *found;
	int i, len, nchoice;

	nchoice = 0;
	choice = calloc((size_t)max, sizeof (char *));

	do {
		if (pt = strpbrk(strval, " \t,")) {
			do {
				*pt++ = '\0';
			} while (strchr(" \t,", *pt));
		}

		if (nchoice >= max) {
			errmsg = TOOMANY;
			return (NULL);
		}
		if (!(menup->attr & CKUNNUM) &&
		    isdigit((unsigned char)*strval)) {
			if (getnum(strval, (int)menup->nchoices, &begin,
			    &end)) {
				errmsg = BADNUM;
				return (NULL);
			}
			chp = menup->choice;
			for (i = 1; chp; i++) {
				if ((i >= begin) && (i <= end)) {
					if (nchoice >= max) {
						errmsg = TOOMANY;
						return (NULL);
					}
					choice[nchoice++] = chp->token;
				}
				chp = chp->next;
			}
			continue;
		}

		found = NULL;
		chp = menup->choice;
		for (i = 0; chp; i++) {
			len = (int)strlen(strval);
			if (strncmp(chp->token, strval, (size_t)len) == 0) {
				if (chp->token[len] == '\0') {
					found = chp->token;
					break;
				} else if (found) {
					errmsg = NOTUNIQ;
					return (NULL); /* not unique */
				}
				found = chp->token;
			}
			chp = chp->next;
		}

		if (menup->invis) {
			for (i = 0; menup->invis[i]; ++i) {
				len = (int)strlen(strval);
				if (strncmp(menup->invis[i], strval,
				    (size_t)len) == 0) {
#if _3b2
					if (chp->token[len] == '\0') {
#else
					if (menup->invis[i][len] == '\0') {
#endif
						found = menup->invis[i];
						break;
					} else if (found) {
						errmsg = NOTUNIQ;
						return (NULL);
					}
					found = menup->invis[i];
				}
			}
		}
		if (found) {
			choice[nchoice++] = found;
			continue;
		}
		errmsg = NULL;
		return (NULL);
	} while (((strval = pt) != NULL) && *pt);
	return (choice);
}

int
setitem(CKMENU *menup, char *choice)
{
	struct _choice_ *chp;
	int n;
	char *pt;

	if (choice == NULL) {
		/* request to clear memory usage */
		chp = menup->choice;
		while (chp) {
			struct _choice_	*_chp = chp;

			chp = chp->next;
			menup->longest = menup->nchoices = 0;

			(void) free(_chp->token); /* free token and text */
			(void) free(_chp);
		}
		return (1);
	}

	if ((chp = calloc(1, sizeof (struct _choice_))) == NULL)
		return (1);

	if ((pt = strdup(choice)) == NULL) {
		free(chp);
		return (1);
	}
	if (!*pt || isspace((unsigned char)*pt)) {
		free(chp);
		return (2);
	}

	chp->token = strtoki(pt, " \t\n");
	chp->text = strtoki(NULL, "");

	if (chp->text) {
		while (isspace((unsigned char)*chp->text))
			chp->text++;
	}
	n = (int)strlen(chp->token);
	if (n > menup->longest)
		menup->longest = (short)n;

	if (insert(chp, menup))
		menup->nchoices++;
	else
		free(chp); /* duplicate entry */
	return (0);
}

int
setinvis(CKMENU *menup, char *choice)
{
	int	index;

	index = 0;
	if (choice == NULL) {
		if (menup->invis == NULL)
			return (0);
		while (menup->invis[index])
			free(menup->invis[index]);
		free(menup->invis);
		return (0);
	}

	if (menup->invis == NULL)
		menup->invis = calloc(2, sizeof (char *));
	else {
		while (menup->invis[index])
			index++; /* count invisible choices */
		menup->invis = realloc(menup->invis,
		    (index+2)* sizeof (char *));
		menup->invis[index+1] = NULL;
	}
	if (!menup->invis)
		return (-1);
	menup->invis[index] = strdup(choice);
	return (0);
}

static int
insert(struct _choice_ *chp, CKMENU *menup)
{
	struct _choice_ *last, *base;
	int n;

	base = menup->choice;
	last = NULL;

	if (!(menup->attr & CKALPHA)) {
		while (base) {
			if (strcmp(base->token, chp->token) == 0)
				return (0);
			last = base;
			base = base->next;
		}
		if (last)
			last->next = chp;
		else
			menup->choice = chp;
		return (1);
	}

	while (base) {
		if ((n = strcmp(base->token, chp->token)) == 0)
			return (0);
		if (n > 0) {
			/* should come before this one */
			break;
		}
		last = base;
		base = base->next;
	}
	if (last) {
		chp->next = last->next;
		last->next = chp;
	} else {
		chp->next = menup->choice;
		menup->choice = chp;
	}
	return (1);
}

void
printmenu(CKMENU *menup)
{
	int i;
	struct _choice_ *chp;
	char *pt;
	char format[16];
	int c;

	(void) fputc('\n', stderr);
	if (menup->label) {
		(void) puttext(stderr, menup->label, 0, 0);
		(void) fputc('\n', stderr);
	}
	(void) sprintf(format, "%%-%ds", menup->longest+5);

	(void) next(NULL);
	chp = ((menup->attr & CKALPHA) ? next(menup->choice) : menup->choice);
	for (i = 1; chp; ++i) {
		if (!(menup->attr & CKUNNUM))
			(void) fprintf(stderr, "%3d  ", i);
		/* LINTED */
		(void) fprintf(stderr, format, chp->token);
		if (chp->text) {
			/* there is text associated with the token */
			pt = chp->text;
			while (*pt) {
				(void) fputc(*pt, stderr);
				if (*pt++ == '\n') {
					if (!(menup->attr & CKUNNUM))
						(void) fprintf(stderr,
						    "%5s", "");
					/* LINTED */
					(void) fprintf(stderr, format, "");
					while (isspace((unsigned char)*pt))
						++pt;
				}
			}
		}
		(void) fputc('\n', stderr);
		chp = ((menup->attr & CKALPHA) ?
		    next(menup->choice) : chp->next);
		if (chp && ((i % 10) == 0)) {
			/* page the choices */
			(void) fprintf(stderr,
			    "\n... %d more menu choices to follow;",
			    menup->nchoices - i);
			(void) fprintf(stderr,
			    /* CSTYLED */
			    "\n<RETURN> for more choices, <CTRL-D> to stop \
display:");
			/* ignore other chars */
			while (((c = getc(stdin)) != EOF) && (c != '\n'))
				;
			(void) fputc('\n', stderr);
			if (c == EOF)
				break; /* stop printing menu */
		}
	}
}

static int
getstr(char *strval, char *defstr, char *error, char *help, char *prompt)
{
	char input[MAX_INPUT];
	char end[MAX_INPUT];

	end[0] = '\0';
	if (defstr) {
		(void) snprintf(end, MAX_INPUT, "(default: %s) ", defstr);
	}
	if (ckquit) {
		(void) strlcat(end, "[?,??,q]", sizeof (end));
	} else {
		(void) strlcat(end, "[?,??]", sizeof (end));
	}

start:
	(void) fputc('\n', stderr);
	(void) puttext(stderr, prompt, 0, 0);
	(void) fprintf(stderr, " %s: ", end);

	if (getinput(input))
		return (1);

	if (strlen(input) == 0) {
		if (defstr) {
			(void) strcpy(strval, defstr);
			return (0);
		}
		puterror(stderr, deferr, (errmsg ? errmsg : error));
		goto start;
	} else if (strcmp(input, "?") == 0) {
		puthelp(stderr, defhlp, help);
		goto start;
	} else if (ckquit && (strcmp(input, "q") == 0)) {
		/* (void) strcpy(strval, input); */
		return (3);
	}
	(void) strcpy(strval, input);
	return (0);
}

static struct _choice_ *
next(struct _choice_ *chp)
{
	static char *last;
	static char *first;
	struct _choice_ *found;

	if (!chp) {
		last = NULL;
		return (NULL);
	}

	found = NULL;
	for (first = NULL; chp; chp = chp->next) {
		if (last && strcmp(last, chp->token) >= 0)
			continue; /* lower than the last one we found */

		if (!first || strcmp(first, chp->token) > 0) {
			first = chp->token;
			found = chp;
		}
	}
	last = first;
	return (found);
}

static char *
strtoki(char *string, char *sepset)
{
	char	*p, *q, *r;
	static char	*savept;

	/* first or subsequent call */
	p = (string == NULL)? savept: string;

	if (p == NULL)		/* return if no tokens remaining */
		return (NULL);

	q = p + strspn(p, sepset);	/* skip leading separators */

	if (*q == '\0')		/* return if no tokens remaining */
		return (NULL);

	if ((r = strpbrk(q, sepset)) == NULL)	/* move past token */
		savept = 0;	/* indicate this is last token */
	else {
		*r = '\0';
		savept = ++r;
	}
	return (q);
}
